polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -2343,64 +2345,50 @@ MLIRASTConsumer::GetOrCreateGlobal(const ValueDecl *FD, std::string prefix,
-      module->getLoc(), builder.getStringAttr(name),
-      /*sym_visibility*/ mlir::StringAttr(), mlir::TypeAttr::get(mr),
-      initial_value, mlir::UnitAttr(), /*alignment*/ nullptr);
-  SymbolTable::setSymbolVisibility(globalOp, lnk);
 
-  globals[name] = std::make_pair(globalOp, isArray);
+  // Initialize the global variable.
+  if (const Expr *Init = Var->getInit()) {
+    MLIRScanner MS(*this, module, LTInfo);
+    mlir::Block B;
+    // In case of device function, we will put the block in the forefront of
+    // the GPU module, else the block will go at the forefront of the main
+    // module.
+    if (FuncContext == FunctionContext::SYCLDevice) {
+      B.moveBefore(getDeviceModule(*module).getBody());
+      MS.getBuilder().setInsertionPointToStart(&B);
+    } else
+      MS.setEntryAndAllocBlock(&B);
 
-  if (tryInit)
-    if (auto init = VD->getInit()) {
-      MLIRScanner ms(*this, module, LTInfo);
-      mlir::Block *B = new Block();
-      // In case of device function, we will put the block in the forefront of
-      // the GPU module, else the block will go at the forefront of the main
-      // module.
-      if (funcContext == FunctionContext::SYCLDevice) {
-        B->moveBefore(getDeviceModule(*module).getBody());
-        ms.getBuilder().setInsertionPointToStart(B);
-      } else
-        ms.setEntryAndAllocBlock(B);
-      OpBuilder builder(module->getContext());
-      builder.setInsertionPointToEnd(B);
-      auto op = builder.create<memref::AllocaOp>(module->getLoc(), mr);
-
-      bool initialized = false;
-      if (isa<InitListExpr>(init)) {
-        if (auto A = ms.InitializeValueByInitListExpr(
-                op, const_cast<clang::Expr *>(init))) {
-          initialized = true;
-          initial_value = A;
-        }
-      } else {
-        auto VC = ms.Visit(const_cast<clang::Expr *>(init));
-        if (!VC.isReference) {
-          if (auto cop = VC.val.getDefiningOp<arith::ConstantOp>()) {
-            initial_value = cop.getValue();
-            initial_value = SplatElementsAttr::get(
-                RankedTensorType::get(mr.getShape(), mr.getElementType()),
-                initial_value);
-            initialized = true;
-          }
-        }
-      }
+    OpBuilder Builder(module->getContext());
+    Builder.setInsertionPointToEnd(&B);
+    auto Op = Builder.create<memref::AllocaOp>(module->getLoc(), VarTy);
 
-      if (!initialized) {
-        FD->dump();
-        init->dump();
-        llvm::errs() << " warning not initializing global: " << name << "\n";
-      } else {
-        globalOp.setInitialValueAttr(initial_value);
-      }
-      delete B;
+    if (isa<InitListExpr>(Init)) {
+      mlir::Attribute InitValAttr =
+          MS.InitializeValueByInitListExpr(Op, const_cast<clang::Expr *>(Init));
+      globalOp.setInitialValueAttr(InitValAttr);
+    } else {
+      bool Initialized = false;
+      ValueCategory VC = MS.Visit(const_cast<clang::Expr *>(Init));
```

Comment:
We still have two `const_cast` here. Can we drop these?

Reply:
I rather not, the reason for the const_cast is that those functions should have been made const to start with and because parts of clang do not observe constant correctness we are forced to either:
   - make other function not be const correct, or
   - inject const_casts 
I prefer the latter.